ArduPlane/ArduPlane.cpp

```diff
@@ -968,6 +969,81 @@ void Plane::update_flight_stage(void)
     airspeed.set_EAS2TAS(barometer.get_EAS2TAS());
 }
 
+/*
+*  ArduSoar support function
+*
+*  Peter Braswell and Samuel Tabor
+*/
+void Plane::update_soaring() {
+    
+    if (!soaring_controller.is_active()) {
+        return;
+    }
+    
+    soaring_controller.update_vario();
+
+    // Check for throttle suppression change.
+    switch (control_mode){
+    case AUTO:
+        soaring_controller.suppress_throttle();
+        break;
+    case FLY_BY_WIRE_B:
+        if (!soaring_controller.suppress_throttle()) {
+            set_mode(RTL, MODE_REASON_SOARING_FBW_B_WITH_MOTOR_RUNNING);
+        }
+        break;
+    case LOITER:
+        // Do nothing. We will switch back to auto/rtl before enabling throttle.
+        break;
+    default:
+        // This does not affect the throttle since suppressed is only checked in the above three modes. 
+        // It ensures that the soaring always starts with throttle suppressed though.
+        soaring_controller.set_throttle_suppressed(true);
+        break;
+    }
+    // Nothing to do if we are in powered flight
+    if (!soaring_controller.get_throttle_suppressed() && aparm.throttle_max > 0) {
+        return;
+    }
+
+    switch (control_mode){
+    case AUTO:
+    case FLY_BY_WIRE_B:
+        // Test for switch into thermalling mode
+        soaring_controller.update_cruising();
+
+        if (soaring_controller.check_thermal_criteria()) {
+            hal.console->printf("Thermal detected, entering loiter\n");
+            set_mode(LOITER, MODE_REASON_SOARING_THERMAL_DETECTED);
+        }
+        break;
+    case LOITER:
+        if (soaring_controller.check_init_thermal_criteria()) {
+            // start by circling around current location
+            soaring_controller.set_wp(next_WP_loc);
+            soaring_controller.update_thermalling(aparm.loiter_radius); // update estimate
```

Comment:
The soaring library can directly access the aparms through the AP_Vehicle layer. This was actually done recently for the landing library exactly for the purpose of fixing the need to pass it around copy it everywhere.